### ISSUE TITLE
Authenticate via Login.gov

### ISSUE THREAD

**Issue by mdragon** [HHS/simpler-grants-gov#2640]
### Summary

Enable grant seekers and applicants to create an account on simpler.grants.gov using the Login.gov single sign-on platform.

This experience will unblock significant parts of our future roadmap, including features like bookmarking opportunities and applying for an opportunity.

Login via other methods (AMS, PIV card) and linking an individual account to an organization are not in scope for this work.

### Press release

We are thrilled to announce that the Simpler Grants.gov effort achieved another milestone today by delivering the ability for users to create a Simpler Grants.gov account. Leveraging the simple, easy-to-use Login.gov sign-in platform, the Simpler Grants.gov account creation flow minimizes the steps and friction users experience when trying to sign up. Supporting account creation paves the way for the team to deliver exciting logged-in features like bookmarking opportunities and applying for an opportunity. Stay tuned for these features and more later this year.

### Acceptance criteria

- [x] Users can create an account on simpler.grants.gov using Login.gov
- [x] Users can log into simpler.grants.gov via Login.gov
- [x] Once logged in, users can see a clear indicator that they are logged in and who they are logged in as
- [x] Users can log out of simpler.grants.gov

### Metrics

- [ ] Number of users: publish a chart showing the number of registered users over time

### Assumptions

This work assumes that we will deliver user value via logged-in features soon. Without other features, user accounts aren't valuable.

### Dependencies

This feature depends heavily on Login.gov. Since Login.gov has a strong operational track record and well-documented APIs, we think the risk involved with this dependency is minimal.

### Original ticket text

Integrate Login.gov with simpler.grants.gov as an authentication service, no additional account creation is required on SGG prior to login/registration. The integration should contain a full round trip from SGG to login.gov and back to SGG with auth token and all expected back/forth fully functioning. Upon successful login the user should be able to signout, and see their name/email in the header (or somewhere else comparable in UI) next to a signout button. Any ‘my account’ page, storage of user info in db, post auth steps, etc. should all be considered separate tickets.

This deliverable will consist of two epics structured such that they could be done in parallel.

**Comment by lucasmbrown-usds** [comment]
Overall LGTM! 

A couple comments: 

1. For the acceptance criteria "Once logged in, users see their email address in the simpler.grants.gov header" - are we sure that's the design we're committing to, or will it possibly change? Maybe we give it slightly more wiggle room, such as "Once logged in, users can see a clear indicator that they are logged in and who they are logged in as, such as seeing their email address in the simpler.grants.gov header." 

2. For Metrics, I like the logged in users count, that's very very simple and straightforward and the essence of the AD. Could we also propose some other metrics as well? Such as: the amount of time it takes for a user to create an account end-to-end, the number of helpdesk tickets we have about account creation on simpler.grants.gov (or some other measure of errors / failures), etc. 

3. Do we want to indicate some kind of stretch goal - so it's not required for the AD, but we'll try to see if we can do it easily - of this kind of variety, all about "Account sharing between Grants.gov and Simpler": 

A. If a user already has an account on Grants.gov, they can login to that same account on simpler.grants.gov to without needing to create a separate account on Simpler.
B. If a new user uses simpler.grants.gov to create a new account, they are able to use Grants.gov to login to the same account without needing to create an account there separately.
C. [This might get moved to a separate Agile Deliverable?] If a user logs into simpler.grants.gov and then navigates to grants.gov, they will remain logged in.

**Comment by mxk0** [comment]
@lucasmbrown-usds 

> For the acceptance criteria "Once logged in, users see their email address in the simpler.grants.gov header" - are we sure that's the design we're committing to, or will it possibly change? Maybe we give it slightly more wiggle room, such as "Once logged in, users can see a clear indicator that they are logged in and who they are logged in as, such as seeing their email address in the simpler.grants.gov header."

Great callout. I just changed the AC to your suggested text.

> For Metrics, I like the logged in users count, that's very very simple and straightforward and the essence of the AD. Could we also propose some other metrics as well? Such as: the amount of time it takes for a user to create an account end-to-end, the number of helpdesk tickets we have about account creation on simpler.grants.gov (or some other measure of errors / failures), etc.

Very open to brainstorming other metrics. I avoided metrics like "amount of time to create an account" because 95% of the flow will happen on Login.gov and is outside of our control. Errors/failures could be a good one to explore further.

> Do we want to indicate some kind of stretch goal - so it's not required for the AD, but we'll try to see if we can do it easily - of this kind of variety, all about "Account sharing between Grants.gov and Simpler":

You suggested this later in your comment, but I think account sharing – assuming the spike indicates it's realistic to do – could end up being its own deliverable. We intentionally scoped this deliverable to be the absolute minimum needed to have accounts working; if we want to set a stretch goal, maybe we set an aggressive timeline (e.g. finish by the end of Quad 1, January 8) rather than increase the scope?

**Comment by lucasmbrown-usds** [comment]
All your comments sound reasonable to me! Thanks. Sounds like next steps are to (1) Explore ideas for error/failure metrics and (2) set a stretch goal on timeline.

**Comment by mxk0** [comment]
Setting a stretch goal to deliver this by the end of Quad 1 (January 8).

**Comment by lucasmbrown-usds** [comment]
We should also start prepping our Security Impact Assessment to note this change in architecture. It should *not* impact any of our controls much because we've already implemented all the appropriate controls for FISMA-Moderate.

**Comment by mxk0** [comment]
## Sprint 1.4 updates

**Status:** 🟢 On track

### Sprint goal

The main goal for this deliverable in Sprint 1.4 was to finalize our auth architecture and create the needed engineering tickets for implementation.

### Accomplishments

- Architecture finalized, eng tickets created

### Rollover

- N/A

### Risks

- We don't see any large risks at this time.

### Sprint 1.5 goals

- This deliverable is prioritized behind finishing our Quad 1 deliverables this sprint
- However, we expect to make progress on both the API and frontend for auth, and should complete the work during Sprint 1.6

**Comment by mdragon** [comment]
## Sprint 1.5 updates

**Status:** 🟢 On track

### Sprint goal

The main goal for this deliverable in Sprint 1.5 was to make progress on both the API and frontend for auth, and should complete the work during Sprint 1.6

### Accomplishments

- Underlying system code built out, following the path laid out in prior Sprints
- Extensive focus on testing and mocking out various parts of the system so we can test and validate both during this initial build and as we iterate, in case Login.gov makes any changes in the future, and in case we need to support additional Auth processes in the future.

### Rollover

- N/A

### Risks

- We don't see any large risks at this time.

### Sprint 1.6 goals

- Developers can login using Login.gov and Simpler understands the user is logged in.

**Comment by lucasmbrown-usds** [comment]
Love that Sprint 1.6 goal!

**Comment by mxk0** [comment]
## Sprint 1.6 and 1.7 updates

**Status:** 🟢 On track

### Sprint goal

The main goal for this deliverable in Sprints 1.6 and 1.7 (which included the holidays and a lot of OoO time) was to complete the auth MVP in a developer branch.

### Accomplishments

- MVP is feature complete

### Rollover

- N/A

### Risks

- We don't see any large risks at this time.

### Sprint 2.1 Goals

- Test the MVP in lower environments so we can coordinate with Login.gov on the move to prod
- Schedule a bug bash with the whole team for Sprint 2.2

**Comment by lucasmbrown-usds** [comment]
> coordinate with Login.gov on the move to prod

There *might* be some paperwork or IAAs to update as we move to PROD, it would be best to figure this out sooner rather than later so we're not blocked. We can discuss with Julius soon or at Wednesday's JAD?

**Comment by mdragon** [comment]
## Sprint 2.1 updates

**Status:** 🟢 On track

### Sprint goal

The main goal for this deliverable in Sprint 2.1 was to have Login.gov working when deployed in the lower Environments.

### Accomplishments
- Test the MVP in lower environments so we can coordinate with Login.gov on the move to Prod
- Schedule a bug bash with the whole team for Sprint 2.2
- Feature flag to allow for early user testing when there's still not a logged in "experience" in Prod.

### Rollover

- N/A

### Risks

- We don't see any large risks at this time.

### Sprint 2.2 Goals
- Finish testing and bug bashing and apply for Prod approval from Login.gov

**Comment by mdragon** [comment]
## Sprint 2.1 updates

**Status:** 🟢 On track

### Sprint goal

Finish testing and bug bashing and apply for Prod approval from Login.gov

### Accomplishments
- Bug bash/testing was completed, no major issues identified.
- Questions to submit for Prod approval were identified

### Rollover
- Compiling answers to complete the Login.gov application and applying for Prod

### Risks

- We don't see any large risks at this time.

### Sprint 2.2 Goals
- Apply for Prod (there's a 2 week turn around so live in Prod would be the following Sprint.

**Comment by mdragon** [comment]
## Sprint 2.3 updates

**Status:** 🟢 On track

### Sprint goal

Apply for Prod (there's a 2 week turn around so live in Prod would be the following Sprint.

### Accomplishments
- Continued follow up on Prod Application for Login.gov questions

### Rollover
- Compiling answers to complete the Login.gov application and applying for Prod
- Apply for Prod (there's a 2 week turn around so live in Prod would be the following Sprint.

### Risks

- We don't see any large risks at this time.

### Sprint 2.4 Goals
Finish identifying application answers and apply for Prod with Login.gov

**Comment by mdragon** [comment]
## Sprint 2.5 updates

**Status:** 🟡  Metrics at risk

### Sprint goal
Finish identifying application answers and apply for Prod with Login.gov

### Accomplishments
- Plan for applying as part of the existing Login.gov Team completed
- Prod App configured in existing Team

### Rollover
- Compiling answers to complete the Login.gov application and applying for Prod
- Apply for Prod (there's a 2 week turn around so live in Prod would be the following Sprint.

### Risks

- Delays in getting Login.gov Prod application approved/submitted put usage Metrics at risk

### Sprint 2.6 Goals
Launch for public Users

**Comment by lucasmbrown-usds** [comment]
Thanks for the great updates. Nitpick: could we uncheck the acceptance criteria for this ticket, since actually the acceptance criteria are not yet met because the work is not available in PROD? I realize the development work is done, and the only hold up is system promotion, but still it would be great to make sure the ACs track the accurate state of the world. 

This will also help making sure our delivery analytics dashboards reflect reality: http://metabase-prod-899895252.us-east-1.elb.amazonaws.com/public/dashboard/94c73a50-033a-4f4a-9383-efd33239bf1d?date=2024-12-30&deliverable=Full+Support+for+Opportunity+Attachments+%28NOFOs%29+%2A&project=13&tab=232-acceptance

**Comment by dustymclean** [comment]
These were the developer notes I **wanted and needed** - thank you.

**Comment by mdragon** [comment]
## Sprint 2.6 updates

**Status:** 🟢  On track

### Sprint goal
- Launch for public Users

### Accomplishments
- Prod implementation tested using Sandbox
- Login.gov App deployed to Prod
- Tested in Prod 🤞🏻 
- Turn on for Public Users 🤞🏻 

### Rollover
- n/a

### Risks

- n/a

### Sprint 2.7 Goals
- Identify metric reporting work for completion in Sprint 2.8

**Comment by mxk0** [comment]
## Sprint 2.7 updates

**Status:** 🟢  Slightly behind

### Sprint goal
- Launch for public Users

### Accomplishments
- Auth live in prod behind a feature flag with one fix for a bug in review

### Rollover
- Turn on for Public Users 

### Risks

- N/A

### Sprint 2.8 Goals
- Async bug bash for auth, saved opportunities, saved search planned for first few days of 2.8
- Publish chart showing # of registered users over time

**Comment by margaretspring** [comment]
@mdragon Can we close this one?? 🎉

**Comment by mdragon** [comment]
## Sprint 2.8 updates

**Status:** 

✅ Users can create an account on simpler.grants.gov using Login.gov
✅ Users can log into simpler.grants.gov via Login.gov
✅ Once logged in, users can see a clear indicator that they are logged in and who they are logged in as
✅ Users can log out of simpler.grants.gov
❌ Number of users: publish a chart showing the number of registered users over time

### Sprint goal
- Deploy fixes for remaining must-fix issues (#3739 and #4242)
- Launch for public Users

### Accomplishments
- Deploy fixes for remaining must-fix issues (#3739 and #4242)
- Launch for public Users

### Rollover
- Publish chart showing # of registered users over time

### Risks

- N/A

### Sprint 3.1 Goals
- Publish chart showing # of registered users over time

**Comment by chrisw-nava** [comment]
I moved the last item (#4615) to a Q3 epic (#4583) for delivery, closing out the remainder of this deliverable!
